### PR TITLE
MINOR: Refactor DelegatingClassLoader to emit immutable PluginScanResult

### PR DESCRIPTION
**Pull request by gharris1727**
The DelegatingClassLoader has a large number of fields and methods for keeping track of known PluginDesc objects. It has this in common with the PluginScanResult data object, which has a similar set of fields and methods. On trunk, the PluginScanResult is only used as a return value of one internal function, before it's results are accumulated by the DCL and discarded. To simplify the DCL, we should use the PluginScanResult object as a container to store and accumulate PluginDesc objects, and return it to the caller to allow them to inspect the scanned plugins without interacting with the DCL.

Using the PluginScanResult as an accumulator, we can delay writing scan results to the pluginLoaders and aliases fields until after all scanning has taken place. This is done via the new installDiscoveredPlugins method. This prevents plugins being scanned from having an inconsistent delegation path: Previously, because the accumulators were updated as scanning proceeded, plugins may or may not be able to see one another via the DelegatingClassLoader::loadClass method. Now, plugins will _not_ be able to see one another before installDiscoveredPlugins is applied and scanning is finished.

This is a first-pass refactor, before pulling the scanning logic out of the DCL to further simplify it. Once external scanning is complete, the caller will call installDiscoveredPlugins to finish initialization of the DCL.

In the trunk implementation, there is some order-sensitivity to the scanning process because of the accumulator fields. In particular: 
1. Because the classpath is scanned last, plugins on the classpath automatically take precedence over isolated plugins. In order to replicate this effect, PluginDesc objects now explicitly compare their isolated/non-isolated nature, and order classpath plugins first.
2. The alias mechanism attempts to determine if aliases are unique, but on trunk it allows colliding aliases from different plugin types to overwrite one another. This is changed to prevent alias collisions across plugin types, now that the iteration over PluginDesc objects isn't grouped by plugin type.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginScanResult.java
Comment (C0urante) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginScanResult.java]: This can be simplified if we tweak the type signature of `allPlugins`:
```suggestion
    public void forEach(Consumer<PluginDesc<?>> consumer) {
        allPlugins.forEach(plugins -> plugins.forEach(consumer));
    }
```

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginScanResult.java
Comment (C0urante) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginScanResult.java]: Since this is private and unlikely to be used elsewhere, we don't have to make it generic (which also makes it trickier to read):
```suggestion
    private static <R extends Comparable<R>> SortedSet<R> merge(List<PluginScanResult> results, Function<PluginScanResult, SortedSet<R>> pluginsAccessor) {
        SortedSet<R> merged = new TreeSet<>();
        for (PluginScanResult element : results) {
            merged.addAll(pluginsAccessor.apply(element));
        }
        return merged;
    }
```

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginScanResult.java
Comment (C0urante) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginScanResult.java]: We can be more specific here, which simplifies the `PluginScanResult::forEach` implementation below:
```suggestion
    private final List<SortedSet<? extends PluginDesc<?>>> allPlugins;
```

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginDesc.java
Comment (C0urante) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginDesc.java]: This getter isn't used anywhere; can we remove it?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginUtils.java
Comment (C0urante) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginUtils.java]: We shouldn't leave this in as a TODO; we should preserve the existing behavior where ambiguous aliases are completely disabled.

**Review by C0urante — COMMENTED** [review]
Thanks Greg. Some minor stylistic comments. Also not sure we should be modifying the aliasing logic without a very good reason; let me know if there's implications that I'm missing.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginUtils.java
Comment (gharris1727) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginUtils.java]: Just to clarify, the code in this PR matches the behavior on trunk, which is that some ambiguous aliases are allowed when they are ambiguous across different plugin loaders.

The TODO is for us to discuss whether to keep the trunk behavior, or change the behavior to drop these ambiguous aliases.

Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginUtils.java]: Ah, my understanding was that we disallowed ambiguous aliases on trunk under all circumstances, except maybe when the same alias is used for two plugins of a different type.

Can you give a specific example of a scenario where we permit ambiguous aliases right now?

Comment (gharris1727) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginUtils.java]: Oh sorry let me clarify, it can produce ambiguous aliases if different plugin types have the same aliases.

Here's an ambiguous alias that can happen on trunk:
```
plugin-1
    package.SomethingConnector
plugin-2
    package.SomethingConverter
```

In this situation, if the iteration scans `plugin-1` first, `Something` will be aliased to `package.SomethingConnector`.
If the iteration scans `plugin-2` first, `Something` will be aliased to `package.SomethingConverter`.

I see that the proposed implementation is actually worse than trunk, which I'll need to improve. I'm still interested if you think that the ambiguity on trunk should be changed or preserved.

Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginUtils.java]: Ah, gotcha--thanks for the clarification 👍

I find the ambiguity that we permit on trunk pretty distasteful, but it's also possible that there are connectors/Kafka Connect clusters out there that are working because of it.

I have no problem with eliminating that ambiguity when using service loader-based plugin discovery.

I have some qualms about eliminating it when using the existing plugin scanning-based discovery. Would it add too much implementation complexity to either preserve the existing behavior (ugly though it is), or at least emit clear error messages in exceptions that get thrown during connector startup and/or config validation?

I think I could live with it if we eliminated ambiguity altogether since the existing behavior feels like the result of an oversight and may be difficult to preserve, but since you're in the neighborhood, it'd be nice to have your thoughts on what (if anything) we can do to reasonably accommodate our users if they're currently relying on it.

Comment (gharris1727) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginUtils.java]: I don't think that anyone should be depending on these cross-plugin ambiguous aliases, because if they were, they would be subject to spurious ClassNotFoundExceptions whenever the non-deterministic iteration order caused a plugin of the wrong type to be loaded.

I've updated this PR to make the failure deterministic, and hopefully easier to debug and remediate. Additionally, I've added a warn log message that explicitly lists the ambiguous aliases and the full class names that they refer to.

An alternative is to separate these aliases into per-plugin-type namespaces, but I don't think that is necessary at this time.

Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginUtils.java]: > I don't think that anyone should be depending on these cross-plugin ambiguous aliases, because if they were, they would be subject to spurious ClassNotFoundExceptions whenever the non-deterministic iteration order caused a plugin of the wrong type to be loaded.

Plugin scanning and alias computation is currently deterministic (the various `Set<PluginDesc<?>>` objects that we use are all instances of the `TreeSet` type), so users are not subject to this risk while running a fixed version of Kafka Connect with a fixed plugin path (and fixed contents of that plugin path).

That said, I think the changes you've made are sufficient.

> An alternative is to separate these aliases into per-plugin-type namespaces, but I don't think that is necessary at this time.

Yeah, this would be nice to have but it's not a blocker. One case where it could come in handy is with plugins that implement multiple plugin interfaces (which is actually fairly common with `Converter` and `HeaderConverter` instances); you may have multiple `JsonConverter` plugins on your worker, but if only one of them implements the `Converter` interface and another implements the `Converter` and `HeaderConverter` interface, it's hard to argue that you shouldn't be able to use an alias when using the latter as a header converter.

**Comment by C0urante** [comment]
Thanks Greg! I think the tradeoffs here are acceptable. Can we add some testing coverage for `PluginUtils::computeAliases`? Once that's done this should be good to merge.

**Comment by gharris1727** [comment]
Here's an example of the new log message, pulled from the test run:
```
[2023-06-02 10:15:57,932] WARN Ambiguous alias 'Colliding' refers to multiple distinct plugins [org.apache.kafka.connect.runtime.isolation.PluginUtilsTest$CollidingHeaderConverter, org.apache.kafka.connect.runtime.isolation.PluginUtilsTest$CollidingConverter]: ignoring (org.apache.kafka.connect.runtime.isolation.PluginUtils:378)

```

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/PluginUtilsTest.java
Comment (C0urante) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/PluginUtilsTest.java]: We can compare entire `Map` objects here, which gives better coverage and improves assertion error messages:
```suggestion
        Map<String, String> actualAliases = PluginUtils.computeAliases(result);

        Map<String, String> expectedAliases = new HashMap<>();
        expectedAliases.put("MockSinkConnector", MockSinkConnector.class.getName());
        expectedAliases.put("MockSink", MockSinkConnector.class.getName());
        expectedAliases.put("MockSourceConnector", MockSourceConnector.class.getName());
        expectedAliases.put("MockSource", MockSourceConnector.class.getName());
        expectedAliases.put("CollidingConverter", CollidingConverter.class.getName());
        expectedAliases.put("Colliding", CollidingConverter.class.getName());
        assertEquals(expectedAliases, actualAliases);
```

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/PluginUtilsTest.java
Comment (C0urante) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/PluginUtilsTest.java]: Nit: double empty line
```suggestion

```

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/PluginUtilsTest.java
Comment (C0urante) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/PluginUtilsTest.java]: Same thought RE comparing whole `Map` objects:

```suggestion
        Map<String, String> actualAliases = PluginUtils.computeAliases(result);

        Map<String, String> expectedAliases = new HashMap<>();
        expectedAliases.put("MockSinkConnector", MockSinkConnector.class.getName());
        expectedAliases.put("MockSink", MockSinkConnector.class.getName());
        assertEquals(expectedAliases, actualAliases);
```

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/PluginUtilsTest.java
Comment (C0urante) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/PluginUtilsTest.java]: Same thought RE comparing whole `Map` objects:

```suggestion
        Map<String, String> actualAliases = PluginUtils.computeAliases(result);

        Map<String, String> expectedAliases = new HashMap<>();
        expectedAliases.put("CollidingConverter", CollidingConverter.class.getName());
        expectedAliases.put("CollidingHeaderConverter", CollidingHeaderConverter.class.getName());
        assertEquals(expectedAliases, actualAliases);
```

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/PluginUtilsTest.java
Comment (C0urante) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/PluginUtilsTest.java]: Same thought RE comparing whole `Map` objects:

```suggestion
        Map<String, String> actualAliases = PluginUtils.computeAliases(result);

        Map<String, String> expectedAliases = new HashMap<>();
        expectedAliases.put("CollidingConverter", CollidingConverter.class.getName());
        assertEquals(expectedAliases, actualAliases);
```

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginUtils.java
Comment (C0urante) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginUtils.java]: After seeing an example of this log message, I'm wondering if the trailing `: ignoring` part here is really beneficial. Maybe something like this would be a bit clearer?

```suggestion
                log.warn("Ignoring ambiguous alias '{}' since it refers to multiple distinct plugins {}", alias, classNames);
```

**Review by C0urante — COMMENTED** [review]
Thanks Greg! Really appreciate the added testing coverage for alias computation. Some small test comments and one logging suggestion and then this should be good to go.

**Review by C0urante — APPROVED** [review]
Thanks Greg! LGTM

Will merge pending CI build

**Comment by C0urante** [comment]
@gharris1727 there are test failures in the latest CI run. Can you look into those? Feel free to just rebase on the latest trunk if that's all that's required.

**Comment by C0urante** [comment]
Test failures appear unrelated (there is a known bug in the reflections library that we use that caused a test failure which may appear related, but is not). Merging...

**Pull request merged**
